sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetime.scala

```diff
@@ -0,0 +1,387 @@
+
+  override protected val cntPerInterval: Int = 24
+
+  override def dataType: DataType = IntegerType
+
+  override def toString: String = s"Hour($child)"
+}
+
+case class Minute(child: Expression) extends TimeFormatExpression {
+
+  override protected val format: Int = 1000 * 60
+
+  override protected val cntPerInterval: Int = 60
+
+  override def dataType: DataType = IntegerType
+
+  override def toString: String = s"Minute($child)"
+}
+
+case class Second(child: Expression) extends TimeFormatExpression {
+
+  override protected val format: Int = 1000
+
+  override protected val cntPerInterval: Int = 60
+
+  override def dataType: DataType = IntegerType
+
+  override def toString: String = s"Second($child)"
+}
+
+abstract class DateFormatExpression extends UnaryExpression with ExpectsInputTypes {
+  self: Product =>
+
+  override def dataType: DataType = IntegerType
+
+  override def expectedChildTypes: Seq[DataType] = Seq(TimestampType)
+
+  protected def defineCodeGen(
+    ctx: CodeGenContext,
+    ev: GeneratedExpressionCode,
+    f: (String, String) => String): String = {
+
+    val tz = classOf[TimeZone].getName
+
+    val utcTime = ctx.freshName("utcTime")
+    val dayInYear = ctx.freshName("dayInYear")
+    val days = ctx.freshName("days")
+    val year = ctx.freshName("year")
+
+    val eval = child.gen(ctx)
+    ev.isNull = eval.isNull
+    eval.code + s"""
+      ${ctx.javaType(dataType)} ${ev.primitive} = ${ctx.defaultValue(dataType)};
+      if (!${ev.isNull}) {
+        long $utcTime = ${eval.primitive} / 10000;
+        long $days = $utcTime / 1000 / 3600 / 24;
+        int $year = (int) ($days / 365.24);
+        int $dayInYear = (int) ($days - $year * 365.24);
+        ${f(dayInYear, utcTime)}
+      }
+    """
+  }
+
+  def eval(input: InternalRow, f: (Int, Long) => Int): Any = {
+    val valueLeft = child.eval(input)
+    if (valueLeft == null) {
+      null
+    } else {
+      val utcTime: Long = valueLeft.asInstanceOf[Long] / 10000
+      val days = utcTime / 1000 / 3600 / 24
+      val year = days / 365.24
+      val dayInYear = days - year.toInt * 365.24
+      f(dayInYear.toInt, utcTime)
+    }
+  }
+
+  override def toString: String = s"Year($child)"
+}
+
+
+case class Year(child: Expression) extends DateFormatExpression {
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+
+    val cal = classOf[Calendar].getName
+    defineCodeGen(ctx, ev, (day, utc) =>
+      s"""
+         if ($day > 1 && $day < 360) {
+           ${ev.primitive} = (int) (1970 + ($utc / 1000 / 3600 / 24 / 365.24));
+         } else {
+           $cal c = $cal.getInstance();
+           c.setTimeInMillis($utc);
+           ${ev.primitive} = c.get($cal.YEAR);
+         }
+       """)
+  }
+
+  override def eval(input: InternalRow): Any = {
+    eval(input, (dayInYear, utcTime) =>
+      if (dayInYear > 1 && dayInYear < 360) {
+        1970 + (utcTime / 1000 / 3600 / 24 / 365.24).toInt
```

Comment:
Is it okay to recalculate the result or should `Year` be independent of `DateFormatExpression`?